File: backend/app/api/v1/endpoints/goals.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from datetime import datetime

from app.database import get_db
from app.models.goal import Goal
from app.models.user import User
from app.schemas.goal import GoalCreate, GoalUpdate, GoalResponse
from app.auth import get_current_user

router = APIRouter()

@router.get("/", response_model=List[GoalResponse])
def get_goals(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    goals = db.query(Goal).filter(Goal.user_id == current_user.id).order_by(Goal.created_at.desc()).offset(skip).limit(limit).all()
    return goals

@router.post("/", response_model=GoalResponse, status_code=status.HTTP_201_CREATED)
def create_goal(
    goal: GoalCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_goal = Goal(
        user_id=current_user.id,
        text=goal.text
    )
    db.add(db_goal)
    db.commit()
    db.refresh(db_goal)
    return db_goal

@router.patch("/{goal_id}", response_model=GoalResponse)
def update_goal(
    goal_id: int,
    goal_update: GoalUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_goal = db.query(Goal).filter(Goal.id == goal_id, Goal.user_id == current_user.id).first()
    if not db_goal:
        raise HTTPException(status_code=404, detail="Goal not found")
        
    db_goal.is_completed = goal_update.is_completed
    if goal_update.is_completed:
        db_goal.completed_at = datetime.utcnow()
    else:
        db_goal.completed_at = None
        
    db.commit()
    db.refresh(db_goal)
    return db_goal

@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_goal(
    goal_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    db_goal = db.query(Goal).filter(Goal.id == goal_id, Goal.user_id == current_user.id).first()
    if not db_goal:
        raise HTTPException(status_code=404, detail="Goal not found")
        
    db.delete(db_goal)
    db.commit()
    return None

File: frontend/src/api/goals.ts
import api from './client';

export interface Goal {
  id: number;
  user_id: number;
  text: string;
  is_completed: boolean;
  created_at: string;
  completed_at?: string;
}

export const goalsAPI = {
  getGoals: async (): Promise<Goal[]> => {
    const response = await api.get('/goals/');
    return response.data as any;
  },
  createGoal: async (text: string): Promise<Goal> => {
    const response = await api.post('/goals/', { text });
    return response.data as any;
  },
  updateGoal: async (id: number, is_completed: boolean): Promise<Goal> => {
    const response = await api.patch(`/goals/${id}`, { is_completed });
    return response.data as any;
  },
  deleteGoal: async (id: number): Promise<void> => {
    await api.delete(`/goals/${id}`);
  }
};
